lib/cog/adapters/hipchat/api.ex

```diff
@@ -1,153 +1,219 @@
+    GenServer.call(__MODULE__, {:retrieve_last_message, room, not_before})
   end
-  def lookup_room(id, as_user: _) do
-    case lookup_user(name: "@" <> id) do
-      %{id: nil} ->
-        case lookup_room(lookup_room(name: id)) do
-          room ->
-            {:ok, room}
-        end
-      user ->
-        lookup_direct_room(user_id: user.id)
+
+  def init(config) do
+    token = config[:api][:token]
+    client = %{token: token}
+
+    case authenticate(client) do
+      :ok ->
+        {:ok, %{client: client}}
+      {:error, :unauthorized} ->
+        raise "Authentication with the HipChat API failed. Please check your api token and try again."
+      {:error, :nxdomain} ->
+        raise "Connecting to the HipChat API failed. Please check your network settings and try again."
     end
   end
 
-  def lookup_room(opts) do
-    lookup("room", option_key(opts))
-    |> Map.get(:body)
-    |> normalize_room
+  def handle_call({:send_message, %{"id" => room_id}, message}, _from, state) do
+    uri = "/room/#{room_id}/message"
+    body = %{message: message}
+    result = post(state.client, uri, body: body)
+    {:reply, result, state}
   end
 
-  def retrieve_last_message(room_name, not_before) do
-    room = lookup_room(name: room_name)
+  def handle_call({:send_message, %{"direct" => user_id}, message}, _from, state) do
+    uri = "/user/#{user_id}/message"
+    body = %{message: message}
+    result = post(state.client, uri, body: body)
+    {:reply, result, state}
+  end
 
-    url = "/room/#{room.id}/history/latest"
-    query = URI.encode_query("max-results": 2, "not-before": not_before)
+  def handle_call({:lookup_room, name: room_name}, _from, state) do
+    uri = "/room/#{room_name}"
 
-    get!(url <> "?" <> query)
-    |> Map.get(:body)
-    |> normalize_last_message
+    result = with {:ok, room} <- get(state.client, uri) do
+      {:ok, normalize_room(room)}
+    end
+
+    {:reply, result, state}
   end
 
-  defp option_key(opts), do: opts[:id] || opts[:handle] || opts[:name]
+  def handle_call({:lookup_user, id: user_id}, _from, state) do
+    uri = "/user/#{user_id}"
+    result = get(state.client, uri)
 
-  defp normalize_user(user)  do
-    %{id: user["id"],
-      handle: user["mention_name"]}
+    result = with {:ok, user} <- result do
+      {:ok, normalize_user(user)}
+    end
+
+    {:reply, result, state}
   end
 
-  defp normalize_room(room) do
-    %{id: room["id"],
-      name: room["name"]}
+  def handle_call(:lookup_users, _from, state) do
+    uri = "/user"
+    result = get(state.client, uri)
+
+    result = with {:ok, %{"items" => users}} <- result do
+      {:ok, Enum.map(users, &normalize_user/1)}
+    end
+
+    {:reply, result, state}
   end
 
-  defp normalize_last_message(%{"items" => [_sent_message, item|_]}),
-    do: item["message"]
-  defp normalize_last_message(%{"items" => _}),
-    do: nil
+  def handle_call({:retrieve_last_message, %{id: room_id}, not_before}, _from, state) do
```

Comment:
This is used in the integration tests. I could probably only include it if `MIX_ENV=test` but I sorta just wanted to replicate the API we already had when converting this over.